cloudbridge/cloud/providers/azure/resources.py

```diff
@@ -0,0 +1,1781 @@
+        self._volume.tags.update(Name=value)
+        self._provider.azure_client. \
+            update_disk_tags(self.id,
+                             self._volume.tags)
+
+    @property
+    def description(self):
+        return self._volume.tags.get('Description', None)
+
+    @description.setter
+    def description(self, value):
+        self._volume.tags.update(Description=value)
+        self._provider.azure_client. \
+            update_disk_tags(self.id,
+                             self._volume.tags)
+
+    @property
+    def size(self):
+        return self._volume.disk_size_gb
+
+    @property
+    def create_time(self):
+        return self._volume.time_created.strftime("%Y-%m-%dT%H:%M:%S.%f")
+
+    @property
+    def zone_id(self):
+        return self._volume.location
+
+    @property
+    def source(self):
+        if self._volume.creation_data.source_uri:
+            url_params = azure_helpers.\
+                parse_url(SNAPSHOT_RESOURCE_ID,
+                          self._volume.creation_data.source_uri)
+            return self._provider.storage.snapshots. \
+                get(url_params.get(SNAPSHOT_NAME))
+        return None
+
+    @property
+    def attachments(self):
+        """
+        Azure does not have option to specify the device name
+        while attaching disk to VM. It is automatically populated
+        and is not returned. As a result this method ignores
+        the device name parameter and passes None
+        to the BaseAttachmentInfo
+        :return:
+        """
+        if self._volume.managed_by:
+            url_params = azure_helpers.parse_url(INSTANCE_RESOURCE_ID,
+                                                 self._volume.managed_by)
+            return BaseAttachmentInfo(self,
+                                      url_params.get(VM_NAME),
+                                      None)
+        else:
+            return None
+
+    def attach(self, instance, device=None):
+        """
+        Attach this volume to an instance.
+        """
+        try:
+            instance_id = instance.id if isinstance(
+                instance,
+                Instance) else instance
+            vm = self._provider.azure_client.get_vm(instance_id)
+
+            vm.storage_profile.data_disks.append({
+                'lun': len(vm.storage_profile.data_disks),
+                'name': self.id,
+                'create_option': 'attach',
+                'managed_disk': {
+                    'id': self.resource_id
+                }
+            })
+            self._provider.azure_client \
+                .create_vm(instance_id, vm)
+            return True
+        except CloudError as cloudError:
+            log.exception(cloudError.message)
+            return False
+
+    def detach(self, force=False):
+        """
+        Detach this volume from an instance.
+        """
+        virtual_machine = None
+        for index, vm in enumerate(
```

Comment:
@afgane I found this quite by luck because the linter complained that the index variable was unused. However, this method has severe issues - it is enumerating through every single vm to detach a disk and secondly, it's creating a vm a few lines below! Looks like we'll need to do a careful inspection of all the code.

Reply:
Ok. Looks like a false alarm for the latter. It's just a misnamed method - it should be create_or_update_vm.

Reply:
We may need to parameterize the detach method with the instance, because Azure doesn't seem to maintain a reference at the disk level: https://docs.microsoft.com/en-us/rest/api/compute/manageddisks/disks/disks-get